doc/developer/design/20230223_stabilize_with_mutually_recursive.md

```diff
@@ -0,0 +1,297 @@
+
+The `WITH MUTUALLY RECURSIVE` (WMR) implementation that was shipped as part of the previous design doc[^wmr] has some loose ends.
+The aim of the design doc is to identify those and come up with a rollout plan for WMR to production environments.
+
+# Motivation
+[motivation]: #motivation
+
+Stabilizing support for `WITH MUTUALLY RECURSIVE` is one of the technical bets that we are making in FY2024.
+Adding first-class support for recursive queries will:
+
+1. Exercise one of the key strengths of the underlying runtime (support for incremental maintenance of iterative dataflows).
+1. Enable new use cases across different domains, most likely based on various forms of graph analysis (for example for social networks, fraud detection, software security).
+1. Enable tractable encodings of high-level concepts such as session windows in terms of SQL (see MaterializeInc/materialize#8698).
+
+We should fill in the implementation gaps that were intentionally left as TODOs during MaterializeInc/materialize#11176 and bring the feature into a shape where it can be gradually rolled out behind a feature toggle and ultimately stabilized.
+
+# Explanation
+[explanation]: #explanation
+<!--
+Explain the design as if it were part of Materialize and you were teaching the team about it.
+This can mean:
+
+- Introduce new named concepts.
+- Explain the feature using examples that demonstrate product-level changes.
+- Explain how it builds on the current architecture.
+- Explain how engineers and users should think about this change, and how it influences how everyone uses the product.
+- If needed, talk though errors, backwards-compatibility, or migration strategies.
+- Discuss how this affects maintainability, or whether it introduces concepts that might be hard to change in the future.
+-->
+
+The original WMR design doc[^wmr] and the technical article[^wmr_article] on our website provide a good explanation about the syntax and semantics of WMR.
+From the implementation point of view, adding WMR support requires extensions across every stage of the query lifecycle.
+The original design doc laid out a plan for adding support in each of the following layers.
+
+1. SQL parsing (&check; in MaterializeInc/materialize#16509)
+2. SQL name resolution (&check; in MaterializeInc/materialize#16509)
+3. SQL planning (&check; in MaterializeInc/materialize#16509)
+4. HIR generalization (&check; in MaterializeInc/materialize#16561)
+5. Lowering (&check; in MaterializeInc/materialize#16561)
+6. MIR generalization (&check; in MaterializeInc/materialize#16561)
+7. MIR optimization corrections (focus of this document)
+8. LIR generalization (&check; in MaterializeInc/materialize#16656, MaterializeInc/materialize#17705)
+9. Rendering (&check; in MaterializeInc/materialize#16787)
+
+The outstanding tasks can be summarized as follows:
+
+1. Complete TODOs that fall under the "MIR optimization corrections".
+1. Enumerate and resolve TODOs in the other stages (if there are any).
```

Comment:
TODO: #16800 seems to be something that needs to be explicitly called out in the "reference explanation".